instrumentation/reactor-3.1/library/src/main/java/io/opentelemetry/instrumentation/reactor/ReactorAsyncSpanEndStrategy.java

```diff
@@ -57,6 +62,10 @@ public EndOnFirstNotificationConsumer(BaseTracer tracer, Context context) {
       accept(null);
     }
 
+    public void onCancel() {
+      accept(new CancellationException());
```

Comment:
The grpc instrumentation adds a "canceled" attribute instead, which seems like a good compromise between failure and success:

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/f755b3f52c5873da01a715bbc805da33e6d0d5d6/instrumentation/grpc-1.6/library/src/main/java/io/opentelemetry/instrumentation/grpc/v1_6/TracingServerInterceptor.java#L130-L141

Unfortunately there isn't a semantic attribute for such a canceled attribute, so we'd need to go with something like:

* `reactor.canceled`
* `rxjava.canceled`

And unfortunately we hide non-semantic attributes behind "experimental" flags (search for `captureExperimentalSpanAttributes` and `CAPTURE_EXPERIMENTAL_SPAN_ATTRIBUTES`), which would probably mean not using singletons for the strategies anymore so that flag can be injected.

Reply:
That approach makes sense and I can change the strategies to adopt that pattern.  It does force a breaking change but it's probably better (and more future-proof) to adopt the same create/builder strategy anyway.  I'll amend the PR.

Reply:
Should we do the same with the Java 8 and Guava strategies?  They do propagate cancellation through completing exceptionally but that is something we can detect and end the span successfully with a similar optional semantic attribute.  It'd be an excuse to change the singleton implementations of those strategies to also follow the create/builder pattern.  I could also submit that as a separate PR if we want to keep this one from getting too complicated.

Reply:
> Should we do the same with the Java 8 and Guava strategies?

If it's possible -- sure, why not.

>  I could also submit that as a separate PR if we want to keep this one from getting too complicated.

💯 👍